src/core/modules/impl/handler/JsHandlerApi.ts

```diff
@@ -54,39 +61,111 @@ export class JsHandlerApi<State> extends AbstractContractHandler<State> {
+    // maybe modify if constructor used
+    this.swGlobal._activeTx = interactionTx;
+    this.swGlobal.caller = interaction.caller; // either contract tx id (for internal writes) or transaction.owner
+  }
+
+  private enableInternalWrites<Input>(
+    executionContext: ExecutionContext<State, unknown>,
+    interactionTx: GQLNodeInterface
+  ) {
+    this.assignReadContractState(executionContext, interactionTx);
+    this.assignViewContractState<Input>(executionContext);
+    this.assignWrite(executionContext);
+    this.assignRefreshState(executionContext);
+  }
+
+  private disableInternalWritesForConstructor() {
+    const templateErrorMessage = (op) =>
+      `Can't ${op} foreign contract state: Internal writes feature is not available in constructor`;
+    this.swGlobal.contracts.readContractState = () =>
+      throwErrorWithName('ConstructorError', templateErrorMessage('readContractState'));
+    this.swGlobal.contracts.write = () => throwErrorWithName('ConstructorError', templateErrorMessage('write'));
+    this.swGlobal.contracts.refreshState = () =>
+      throwErrorWithName('ConstructorError', templateErrorMessage('refreshState'));
+    this.swGlobal.contracts.viewContractState = () =>
+      throwErrorWithName('ConstructorError', templateErrorMessage('viewContractState'));
+  }
+
+  async handle<Input, Result>(
+    executionContext: ExecutionContext<State>,
+    currentResult: EvalStateResult<State>,
+    interactionData: InteractionData<Input>
+  ): Promise<InteractionResult<State, Result>> {
+    const { interaction, interactionTx } = interactionData;
+
+    this.setupSwGlobal(interactionData);
+    this.enableInternalWrites(executionContext, interactionTx);
+
+    this.assertNotConstructorCall<Input, Result>(interaction);
+
+    return await this.runContractFunction(executionContext, interaction, currentResult.state);
+  }
+
+  private assertNotConstructorCall<Input, Result>(interaction: ContractInteraction<Input>) {
+    if (this.contractDefinition.manifest?.evaluationOptions.useConstructor &&
+      interaction.input['function'] === INIT_FUNC_NAME) {
+      throw new Error(`You have enabled {useConstructor: true} option, so you can't call function ${INIT_FUNC_NAME}`);
     }
   }
 
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  initState(state: State): void {
-    // nth to do in this impl...
+  // eslint-disable-next-line
+  initState(state: State): void { }
+
+  async maybeCallStateConstructor<Input>(
+    initialState: State,
+    executionContext: ExecutionContext<State>
+  ): Promise<State> {
+    if (this.contractDefinition.manifest?.evaluationOptions.useConstructor) {
+      const interaction = {
+        input: { function: INIT_FUNC_NAME, args: initialState } as Input,
+        caller: this.contractDefinition.owner
+      };
+      const interactionTx = { ...this.contractDefinition.contractTx, sortKey: ZERO_SORT_KEY };
```

Comment:
Zero sort key hack is for enabling KV to work in __init

Reply:
so all the values set in KV in the constructor will have the 'zero' sortKey set, right?

Reply:
Right